### PR TITLE
Avoid reordering tables

### PR DESCRIPTION
**Pull request by alsuren**
Adds Document::to_string_in_original_order()

This is intended to help with https://github.com/killercup/cargo-edit/issues/218
    
We save .position = Some(i) on each Table, starting with i = 0 for
the root of the Document, and then incrementing each time we see a
table header.

When serialising, ~we start looking for a Table with position 0,
which we should find at the root of the document. We then continue
scanning, looking for position 1 and so on. If the document doesn't
need re-ordering then we will on need to do a single scan.~ we start by 
doing a walk over the tree and working out which positions exist
(storing them in a Vec and then sorting them). We then do subsequent
walks to find each of the known positions in turn. Each walk of the tree
is guaranteed to yield at least one table, but will probably yield more.
If the document doesn't need re-ordering then we will on need to do a single scan.

If a Table has been added using the API rather than parsed then it
will have position = None. We only want to serialise each Table
once, so we only print tables with position = None if the table
we visited immediately before it had the right position.

If we didn't manage to serialise all Tables after a single loop
then we keep looping until we are done. ~There is code to make sure
we don't get stuck when someone causes a gap in the numbers by
deleting a Table.~ It is impossible for the loop to get stuck, because
it only searches for position values that it already knows are in the tree.

Known issues:
- [x] ~If you have created your Document by parsing two .toml files and
  merging the results together, this method may silently skip some
  tables. Please use Document.to_string() instead, which doesn't have
  this problem.~ fixed
- [ ] The best case performance of this function is a bit worse than
  Document.to_string(). It will be significantly worse if you have lots of
  tables that are in strange orders.

Also adds `pretty_assertions` in some tests and color-backtrace (as separate commits) because I found debugging test failures more tricky without them.

Things to do before merging:
- [x] make a version of cargo-edit which uses this patch, and make sure that it actually works on my test samples at https://github.com/alsuren/rust-repos/tree/cargo.toml-samples
  - see https://github.com/killercup/cargo-edit/pull/312
  - went from 1457 funky files down to 13 (and I think that some of them are because I didn't clean up the unix/windows line endings correctly)
- [x] Write a few more integration tests, describing interesting edge cases from the repo. For example:
  - [x] An example with trailing comments after the end of the document.
  - [x] An example where `[package.metadata.deb]` is at the bottom of the file and it stays there.
    * The table_reordering test already does this.
  - [x] A test with the following document, where you add `[dependencies.somethingelse]` and it is added after `[dependencies.opencl]` rather than before `[[example]]`.
```
[package]
[dependencies]
[[example]]
[dependencies.opencl]
[dev-dependencies]
```
- [x] Is the API okay? Display.to_string() seems to swallow the error case (possibly by panicking). Should we do the same? (answer: yes)
- [x] (edit: added a commit to use this new algorithm) Is it enough to document the issue where it can silently skip tables that are parsed from multiple files, or should I check for it? If I want to fix it properly then the best way to do it is to:
  * walk the tree once, gathering all of the known `position` numbers unto a vec
  * Sort the positions vec
  * Walk the tree and pop off items from the positions vec as I go (rather than incrementing an integer)
  * Stop looping when the vec is empty.
  * This is probably simpler than my current implementation, but forces me to allocate another vec, and then do at least two passes of the tree, even if the tables are in the correct order already.
- [x] update the limitations section of the readme to explain that you should use Document::to_string_in_original_order() if you care about not forcing nested tables to be grouped next to their parents.
  - [ ] Once this PR is merged, make another PR to point `https://github.com/ordian/toml_edit/blob/f09bd5d075fdb7d2ef8d9bb3270a34506c276753/tests/test_valid.rs#L84` at the updated test.

### CODE REVIEW

File: tests/test_edit.rs [hunk 1]
```diff
@@ -82,6 +118,50 @@ dc = "eqdc10"
...
+        "[a]"
+    ).running(|root| {
+        let other = "[b]".parse::<Document>().unwrap();
+        root["b"] = other["b"].clone();
+    }).produces_display(
+        "[a]\n[b]\n"
+    ).produces_in_original_order(
+        // Known limitation of the current approach: b goes missing when
+        // printing.
+        "[a]\n"
```
Comment (alsuren) [inline tests/test_edit.rs]: In this test, `[a]` and `[b]` both have `.position = 1` so only `[a]` gets printed.

Is this enough of a bug that we should switch to the slower re-ordering scheme that I described in the PR description, or is it considered an unsupported edge case?

File: src/table.rs [hunk 2]
```diff
@@ -15,6 +15,9 @@ pub struct Table {
     pub(crate) decor: Decor,
     // whether to hide an empty table
     pub(crate) implicit: bool,
+    // used for putting tables back in their original order when serialising.
+    // Will be None when the Table wasn't parsed from a file.
+    pub(crate) position: Option<usize>,
```
Comment (alsuren) [inline src/table.rs]: The `usize` can actually be anything that implements PartialOrd, if you have a preference.

File: src/display.rs [hunk 3]
```diff
@@ -147,6 +201,29 @@ impl Display for Table {
...
 
+impl Document {
+    /// Returns a string representation of the TOML document, attempting to keep
+    /// the table headers in their original order.
+    ///
+    /// The best case performance of this function is slightly slower than
+    /// Document.to_string(). If you have lots of tables that are in strange
+    /// orders then it may be significantly slower as it has to walk the tree
+    /// multiple times.
+    pub fn to_string_in_original_order(&self) -> std::result::Result<String, Error> {
```
Comment (alsuren) [inline src/display.rs]: The biggest open question is whether I should make the return signature for this method be like Display::to_string() by calling .unwrap() or something.

If you have suggestions for better names for anything then please say. If you think it would be an acceptable change for to_string() to implement this algorithm then we don't need to worry about naming or return types ;-) .

**Comment by ordian** [comment]
I'll try to have look at the PR this week (sorry for the delay, I'm a bit busy lately).

File: src/display.rs (hunk 3 above)
Comment (ordian) [reply src/display.rs]: I don't really understand conditions under which it can return an error. AFAIK, `write!` to string always succeeds, unless we are out of memory, in which case we can't do much about it.

File: src/table.rs (hunk 2 above)
Comment (ordian) [reply src/table.rs]: `usize` seems fine to me.

I'm thinking of weird ways of abusing this field. What happens if we have two documents and we `mem::swap` tables from different documents with different indices? We could end up with two table with the same position.

Comment (ordian) [reply src/table.rs]: of even swappin a new table with an existing table in the document would change the order, right?

Comment (alsuren) [reply src/table.rs]: If we end up with two tables in the same position then we will print them one after the other (a previous implementation in this pr would silently drop them)

If you mem::swap two tables from the same doc like the implicit tables a and z in here:
```
[a.b]
[z.x]
```
Then Display will format it like this because it cares most about the order of entries in the top level Document
```
[a.x]
[z.b]
```
in_original_order will print:
```
[z.b]
[a.x]
```
because it skips the implicit tables and puts b and x in their original order.

This is a bit of a contrived example though.

The more interesting case is if you parse two Documents and then try to append the top level keys from the second doc onto the first. In the current implementation, in_original_order will interleave the tables from each doc, so I would recommend using Display::to_string() instead. I'm not sure it's well defined which order you want though, because both functions will hoist non-Table values to the top. If you don't want interleaving then you could make position be a tuple of (document_id, position), where document_id is an integer that increments whenever you start parsing a document. I'm not sure it's worth the added complexity though, because I don't think anyone will ever want to do this and also care about honouring split tables.

I suppose that any "this is an edge case" argument I make is weakened by the fact that this entire pr is about fixing a 2% edge case, so I will take your lead on this.

Comment (alsuren) [reply src/table.rs]: If we end up with two tables in the same position then we will print them one after the other (a previous implementation in this pr would silently drop them)

If you mem::swap two tables from the same doc like the implicit tables a and z in here:
```
[a.b]
[z.x]
```
Then Display will format it like this because it cares most about the order of entries in the top level Document
```
[a.x]
[z.b]
```
in_original_order will print:
```
[z.b]
[a.x]
```
because it skips the implicit tables and puts b and x in their original order.

This is a bit of a contrived example though.

The more interesting case is if you parse two Documents and then try to append the top level keys from the second doc onto the first. In the current implementation, in_original_order will interleave the tables from each doc, so I would recommend using Display::to_string() instead. I'm not sure it's well defined which order you want though, because both functions will hoist non-Table values to the top. If you don't want interleaving then you could make position be a tuple of (document_id, position), where document_id is an integer that increments whenever you start parsing a document. I'm not sure it's worth the added complexity though, because I don't think anyone will ever want to do this and also care about honouring split tables.

I suppose that any "this is an edge case" argument I make is weakened by the fact that this entire pr is about fixing a 2% edge case, so I will take your lead on this.

File: src/display.rs (hunk 3 above)
Comment (alsuren) [reply src/display.rs]: Okay. I will probably unwrap the error then.

Might not be until the weekend though. Thanks for the advice.

**Comment by alsuren** [comment]
I just pushed the unwrap change and a commit to update the limitations part of the README. I'm happy with the shape of the API at this point. I will want to clean up the commit history once we're happy with the code.

File: src/display.rs [hunk 4]
```diff
@@ -78,6 +78,65 @@ impl Display for InlineTable {
...
+        // it based on whether we printed the previous table we visited. This
+        // is to avoid printing tables more than once.
+        //
+        // We set should_print to true the first time we went around the
+        // loop, so initially None-positioned Tables will have already been
+        // printed.
+        let mut should_print = true;
+
+        let mut current_position: Option<&usize> = position_iter.next();
+        while current_position.is_some() {
```
Comment (ordian) [inline src/display.rs]: The complexity of this algorithm is quadratic in the worst case. I think we can do better here. We could serialize the document to `Vec<(u64, String)>` in one pass of `visit_nested_tables`. Each entry of this vec would represent a table. The `u64` will be an id of the table sorted as in the final `String`, and the second part of the tuple is the `String` representation of the table. Then all we need to do is to sort the vec by id and concatenate the strings.

How do we construct this vec? Something like this:
```rust
       // this number should be at least the number of tables with position `None`
       // alternatively, this could be computed in one pass of `visit_nested_tables`
       const MULTIPLIER: usize = 100500;
       // this would account for all `None` positions
       let mut increment = 0;
       let mut last_position = 0;
       let mut tables = Vec::new();
       self.visit_nested_tables(path, is_array_of_tables, &mut |t, p, is_array| {
            if let Some(pos) = t.position {
                last_position = pos;
                increment = 0;
            } else {
                increment += 1;
            }
            let id = last_position * MULTIPLIER + increment;
            let mut s = String::new();
            visit_table(&mut s, t, p, is_array);
            tables.push((id, s));
            Ok(())
        })?;
        tables.sort_by_key(|&(id, _)| id);
```
This wouldn't work for a generic callback though, so it would require changing the type signature.

Comment (ordian) [reply src/display.rs]: Even better, since `sort` is stable sort, we can simplify the algorithm:
```rust
       let mut last_position = 0;
       let mut tables = Vec::new();
       self.visit_nested_tables(path, is_array_of_tables, &mut |t, p, is_array| {
            if let Some(pos) = t.position {
                last_position = pos;
            }
            let mut s = String::new();
            visit_table(&mut s, t, p, is_array);
            tables.push((last_position, s));
            Ok(())
        })?;
        tables.sort_by_key(|&(id, _)| id);
```

Comment (alsuren) [reply src/display.rs]: Aha! For some reason I didn't think about sorting tuples of position and payload (even though that's exactly what I would do in python). Thanks for the tip.

I'm wondering if there is a payload that we could use that avoids the extra string copy in the happy case, but a Table can't serialise itself without knowing it's path, so String is probably the best payload.

I'm away from the computer this weekend, but I will try to code this up during the week.
